src/legacy/validation_zerocoin_legacy.cpp

```diff
@@ -46,24 +46,24 @@ bool DisconnectZerocoinTx(const CTransaction& tx)
 
 // Legacy Zerocoin DB: used for performance during IBD
 // (between Zerocoin_Block_V2_Start and Zerocoin_Block_Last_Checkpoint)
-void DataBaseAccChecksum(const CBlockIndex* pindex, bool fWrite)
+void CacheAccChecksum(const CBlockIndex* pindex, bool fWrite)
 {
     const Consensus::Params& consensus = Params().GetConsensus();
-    if (!pindex ||
+    if (!pindex || accumulatorCache == nullptr ||
         !consensus.NetworkUpgradeActive(pindex->nHeight, Consensus::UPGRADE_ZC_V2) ||
         pindex->nHeight > consensus.height_last_ZC_AccumCheckpoint ||
         pindex->nAccumulatorCheckpoint == pindex->pprev->nAccumulatorCheckpoint)
         return;
 
     arith_uint256 accCurr = UintToArith256(pindex->nAccumulatorCheckpoint);
     arith_uint256 accPrev = UintToArith256(pindex->pprev->nAccumulatorCheckpoint);
-    // add/remove changed checksums to/from DB
+    // add/remove changed checksums to/from cache
     for (int i = (int)libzerocoin::zerocoinDenomList.size()-1; i >= 0; i--) {
         const uint32_t nChecksum = accCurr.Get32();
         if (nChecksum != accPrev.Get32()) {
             fWrite ?
-            zerocoinDB->WriteAccChecksum(nChecksum, libzerocoin::zerocoinDenomList[i], pindex->nHeight) :
-            zerocoinDB->EraseAccChecksum(nChecksum, libzerocoin::zerocoinDenomList[i]);
+            accumulatorCache->Set(nChecksum, libzerocoin::zerocoinDenomList[i], pindex->nHeight) :
+            accumulatorCache->Erase(nChecksum, libzerocoin::zerocoinDenomList[i]);
```

Comment:
Another dangerous issue from this ancient code that surely caused sync problems and forking situations in the past:

As the accumulators checksums are stored in the key->value db under key 'checksum|denom' and value 'chainHeight' . At the block connection time the `DataBaseAccChecksum` function only store the first block with a different accumulator checkpoint (any subsequent block that has the same `accCheckpoint` is not stored). And, at disconnection time, as it goes in the opposite chain direction, it removes the acc checksums from db of the first block with a different acc checkpoint (the one that stored the value) using the **last** connected block with the same acc checkpoint (the block being disconnected).
Leaving a range of blocks with no databased acc checksums.